src/storage/page_buffer.c

```diff
@@ -1967,37 +1972,59 @@ pgbuf_fix_release (THREAD_ENTRY * thread_p, const VPID * vpid, PAGE_FETCH_MODE f
 	   * page. */
 	}
     }
-  else if (fetch_mode == OLD_PAGE_IF_IN_BUFFER)
-    {
-      /* we don't need to fix page */
-      pthread_mutex_unlock (&hash_anchor->hash_mutex);
-      return NULL;
-    }
   else
     {
-      bufptr = pgbuf_claim_bcb_for_fix (thread_p, vpid, fetch_mode, hash_anchor, &perf, &retry);
-      if (bufptr == NULL)
+      /* The page is not found in the hash chain the caller is holding hash_anchor->hash_mutex */
+
+      if (er_errid () != NO_ERROR)
 	{
-	  if (retry)
-	    {
-	      retry = false;
-	      goto try_again;
-	    }
-	  ASSERT_ERROR ();
-	  return NULL;
+	  // error searching the hash chain
+	  pthread_mutex_unlock (&hash_anchor->hash_mutex);
+	  PGBUF_BCB_CHECK_MUTEX_LEAKS ();
+	  return nullptr;
 	}
-      buf_lock_acquired = true;
 
-#if defined(ENABLE_SYSTEMTAP)
-      if (fetch_mode == NEW_PAGE && pgbuf_hit == false)
+      if (fetch_mode == OLD_PAGE_IF_IN_BUFFER)
 	{
-	  pgbuf_hit = true;
+	  /* we don't need to fix page */
+	  pthread_mutex_unlock (&hash_anchor->hash_mutex);
+	  return NULL;
 	}
-      if (fetch_mode != NEW_PAGE)
+      else
 	{
-	  CUBRID_PGBUF_MISS ();
-	}
+	  /* In this case, the caller is holding only hash_anchor->hash_mutex. The hash_anchor->hash_mutex is to be
+	   * released in pgbuf_lock_page (). */
+	  const PGBUF_LOCK_MODE lock_page_res = pgbuf_lock_page (thread_p, vpid, fetch_mode, hash_anchor, &perf);
+	  if (lock_page_res == PGBUF_LOCK_WAITER)
+	    {
+	      goto try_again;
```

Comment:
I think an explanation here would help:
```suggestion
              // Somebody else loaded page from disk. Try to search the hash again.
	      goto try_again;
```